RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,440 @@
+            // Pointer to the beginning of the current row in RowMajor order
+            T * row_buf = buf + row * num_cols;
+
+            // Apply the original FHT on this row
+            for (int64_t i = 0; i < log_n; ++i) {
+                int64_t s1 = 1 << i;
+                int64_t s2 = s1 << 1;
+                for (int64_t j = 0; j < n; j += s2) {
+                    for (int64_t k = 0; k < s1; ++k) {
+                        // For implicitly padding the input we just have to make sure
+                        // we replace all out-of-bounds accesses with zeros
+                        bool b1 = j + k < num_cols;
+                        bool b2 = j + k + s1 < num_cols;
+                        T u = b1 ? row_buf[j + k] : 0;
+                        T v = b2 ? row_buf[j + k + s1] : 0;
+                        if(b1 && b2) {
+                            row_buf[j + k] = u + v;
+                            row_buf[j + k + s1] = u - v;
+                        }
+                        else if(!b2 && b1) {
+                            row_buf[j + k] = u + v;
+                        }
+                        else if(!b2 && !b1)
+                            continue;
+                    }
+                }
+            }
+        }
+    }
+
+    template <typename T>
+    void fht_right_col_major(T *buf, int64_t log_n, int64_t num_rows, int64_t num_cols) {
+        int64_t n = 1 << log_n;
+
+        // Apply FHT to each row independently
+        for (int64_t row= 0; row < num_rows; ++row) {
+            // Apply the original FHT on this column
+            for (int64_t i = 0; i < log_n; ++i) {
+                int64_t s1 = 1 << i;
+                int64_t s2 = s1 << 1;
+                for (int64_t j = 0; j < n; j += s2) {
+                    for (int64_t k = 0; k < s1; ++k) {
+                        // For implicitly padding the input we just have to make sure
+                        // we replace all out-of-bounds accesses with zeros
+                        bool b1 = j + k < num_cols;
+                        bool b2 = j + k + s1 < num_cols;
+                        T u = b1 ? buf[(j + k) * num_rows + row] : 0;
+                        T v = b2 ? buf[(j + k + s1) * num_rows + row] : 0;
+                        if(b1 && b2) {
+                            buf[(j + k) * num_rows + row] = u + v;
+                            buf[(j + k + s1) * num_rows + row] = u - v;
+                        }
+                        else if(!b2 && b1) {
+                            buf[(j + k) * num_rows + row] = u + v;
+                        }
+                        else if(!b2 && !b1)
+                            continue;
+                    }
+                }
+            }
+        }
+    }
+
+    template <typename T>
+    void fht_dispatch(
+        bool left,
+        blas::Layout layout,
+        T* buff,
+        int64_t log_n,
+        int64_t num_rows,
+        int64_t num_cols
+        )
+    {
```

Comment:
Done! (though I do require access to an `lda` argument since the value of `lda` is apparent from the way I deal with the application of the hadamard transform on a case-by-case basis)